### DISCUSSION TITLE
Query logs in psycopg3 - async connection/cursor

### DISCUSSION THREAD

**Discussion by alairock** [psycopg/psycopg#184] [Q&A]
In psycopg2 there was a LoggingCursor class. I would like to be able to log queries in Psycopg3's async connection/cursor. 

I'm not seeing documentation for enabling query logs in the psycopg3 docs. Anyone know how to do this? Is this still a feature?

**Comment by dvarrazzo** [comment]
There is currently no feature to log every single query. We could introduce them, if there is demand, and a clear use case and design of the feature.

If you would like to prototype what you need you might write a `Cursor` subclass and override, for instance, `_execute_send()`

**Reply by alairock** [reply]
I mean, the clearest use-case is debugging. I have a really complex series of transactions and sub-transactions that I want to confirm are executing in the correct order. It would be much easier to reason about if I could see the raw queries logged to stdout.

**Reply by dvarrazzo** [reply]
Yes, but:

- query and parameters are separated in psycopg 3. What to log? What if a parameter is binary? What if a parameter is 8mb long or a query is 128k long?
- on which logger to log?
- Do we really want to have a separate class or wouldn't rather have logging built-in in the normal `Cursor`? what overhead does logging add? Does it block on async query?
- what use do you make of multiline queries in the logs?
- In a multithread environment, are you sure that the order you see the log is the same of the order that the queries are sent to the database? Wouldn't it better to log on the server, by setting `log_min_duration_statement` to 0?

All in all, the main problem here is that, personally, I have never used the logging cursor for debugging, in 10+ years using psycopg2. As such, introducing it doesn't mean that I would introduce anything useful: for me it would be a tick box exercise and add on my maintenance burden. So, I would rather have a reference implementation of someone who has used query logging and can show an interesting use for it.

**Reply by ddurham2** [reply]
it plainly existed in psycopg2.extras and I use it regularly for this purpose.  Not everyone's situation hits all those complexities.  It's good at times to be able to see the queries that are executing inline with the rest of debug logs.

To your points:
- log the query and the parameters on the same or separate log lines, go parameter by parameter and handle each type and its possible length with some sane decisions. (e.g. strings can be truncated to some number of chars, and binary can be be replaced with "<bin>")  This will catch a high percentage of what's needed. 
- a logger supplied to the LoggingConnection or a default
- A separate class?  That's the easiest thing to do is create a sub-class of Cursor.  It's a slight overhead, but the dev can ask for it or not.  And when they don't, nothing has been affected.  Does it block on async?  no differently than if you log in any async function.
- lines could be joined.. or just let it log as-as would be fine IMO since the logger's formatter could be left to make that decision.
- IMO, in this use-case, you want to see the real order the client issued queries in context of what other logging the client is doing.  One is certainly welcome to watch the server side instead, in addition, or not at all.

None of these issues are insurmountable.  And v2 had dealt with them.
